### ISSUE TITLE
Very slow View on large dataset

### ISSUE THREAD

**Issue by anatolix** [rstudio/rstudio#1705]
Some people use RStudio with quite large data loaded to R. For example you could have data.table containing 100Gb of data. Accidentally clicking View on this data.table hangs up RStudio or RStudio server. 

The reason is simple R will transfer all 100Gb of data to r-studio (it could take 20-30 minutes or even hours if you computer have no additional 100Gb) after it RStudio cuts first 1000 lines of data and displays them. 

The fix for that is very simple, we need cut 1000 lines in R and not transfer everything else to RStidio, I have proof of concept package which fixes this if loaded https://github.com/anatolix/RStudioViewFix/blob/master/R/RStudioViewFix.R
But I think this should be fixed inside R-Studio companion R package which redeclare View function for R-Studio.

How to reproduce bug:
Any platform, Any R-Studio version, Any R-Version.
Load or generate data.frame or data.table 1GB size
Click View (it will not hang up completely but you notice View will be opened not instantly but after several second)

**Comment by kevinushey** [comment]
Here's a reproducible example:

```
big <- data.frame(x = 1:1E8)
View(big)
```

**Comment by kevinushey** [comment]
This particular routine is very slow:

https://github.com/rstudio/rstudio/blob/1867b7d59a747f0bebda18a9e1f9f8da83c083a5/src/cpp/session/modules/SessionDataViewer.R#L179-L183

This is because, when the object has no row names set, R must generate the requested row names on demand. We end up generating 1E8 row names, and then subsetting based on the portion we actually care about. We can detect this name by calling `.row_names_info(x, 0)`:

```
> .row_names_info(big, 0)
[1]         NA -100000000
```

This is the so-called 'compact' row names representation. (Note that the second element is not always negative, but the first element is always NA, and it's always an integer vector.)

In that routine, if we detect that no row names exist, we should manually create the row names for the range we desired, to avoid generating an overly-large row names vector.

**Comment by dfalty** [comment]
Verified fixed 1.2.359
